tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java

```diff
@@ -0,0 +1,294 @@
+        if (!running.compareAndSet(true, false)) {
+            throw new IllegalStateException("ExternalCommandWorker is not running.");
+        }
+        log.info("{}: Deactivating ExternalCommandWorker.", id);
+        if (ep != null) {
+            ep.destroyProcess();
+        }
+        executor.shutdownNow();
+        executor.awaitTermination(1, TimeUnit.DAYS);
+        this.executor = null;
+        this.ep = null;
+        this.status = null;
+        this.doneFuture = null;
+        this.ep = null;
+    }
+
+    public class ControlCommand {
+        private final String action;
+        private final ExternalCommandSpec spec;
+        @JsonCreator
+        ControlCommand(@JsonProperty("action") String action,
+                       @JsonProperty("spec") ExternalCommandSpec spec) {
+            this.action = action;
+            this.spec = spec;
+        }
+        @JsonProperty
+        public String action() {
+            return action;
+        }
+        @JsonProperty
+        public ExternalCommandSpec spec() {
+            return spec;
+        }
+
+        public void issue(PrintWriter controlChannel) {
+            String command = JsonUtil.toJsonString(this);
+            log.info("{}: Send out command: {} ", id, command);
+            controlChannel.println(command);
+        }
+    }
+
+    public class ExternalProcess {
+        private final Process process;
+        private final PrintWriter controlChannel;
+        Future stdoutMonitor;
+        Future stderrMonitor;
+        ExternalProcess() throws IOException {
+            ProcessBuilder processBuilder = new ProcessBuilder(spec.command());
+            this.process = processBuilder.start();
+            this.controlChannel = new PrintWriter(new OutputStreamWriter(process.getOutputStream(), StandardCharsets.UTF_8), true);
+            this.stdoutMonitor = executor.submit(new StatusUpdater(this.process));
+            this.stderrMonitor = executor.submit(new ErrorUpdater(this.process));
+        }
+
+        void startTask() {
+            synchronized (this) {
+                new ControlCommand("start", spec).issue(controlChannel);
+            }
+        }
+
+        int waitForProcess() throws InterruptedException, ExecutionException {
+            process.waitFor();
+            stdoutMonitor.get();
+            stderrMonitor.get();
+            return process.exitValue();
+        }
+        void stopTask() {
+            synchronized (this) {
+                new ControlCommand("stop", spec).issue(controlChannel);
+            }
+        }
+        void destroyProcess() throws InterruptedException {
+            if (process.isAlive()) {
+                this.stopTask();
+                process.waitFor(1, TimeUnit.MINUTES);
+                if (process.isAlive()) {
+                    process.destroy();
+                    process.waitFor();
+                }
+            }
+        }
+    }
+
+    public class ErrorUpdater implements Runnable {
```

Comment:
Let's call this "StderrMonitor" since that is a better description of what it does.  As I commented earlier, stderrMonitor is what the instance variable is called, which again suggests that stderrMonitor is a more descriptive name.

Reply:
Done